### ISSUE TITLE
Allow dynamic default value for date fields

### ISSUE THREAD

**Issue by dwt** [lektor/lektor#883]
I would like to say something like this in my blog post model:

```ini
[fields.pub_date]
label = Publication date
type = date
width = 1/4
default = {{ date.today() }}
```

Or something that acomplishes something like that. I just had a quick look into how that would be implemented, and a fairly generic patch like this would totally do it:

```diff
diff --git a/lektor/datamodel.py b/lektor/datamodel.py
index c8ff39a..ceb1355 100644
--- a/lektor/datamodel.py
+++ b/lektor/datamodel.py
@@ -179,6 +179,8 @@ class Field:
         self.label_i18n = label_i18n
         self.description_i18n = get_i18n_block(options, "description") or None
         self.default = options.get("default")
+        if self.default and self.default.startswith('{{') and self.default.endswith('}}'):
+            self.default = FormatExpression(env, self.default).evaluate()
         self.type = type(env, options)
 
     @property
```

However that seems quite hackisch. What do you guys think about this - and where / how would one implement something like this cleanly?

**Comment by dwt** [comment]
My first idea was to implement this in `lektor.types.primitives.DateType()`, but that doesn't seem to be possible, as it is actually the `lektor.datamodel.Field()` that keeps track of the default value, thus the `DateType()` can't seem to change how the default is computed?

**Comment by dairiki** [comment]
Types can change how they handle default values by overriding `Type.value_from_raw_with_default`. 

----

I'm not sure how useful this is, however.  Note that your pub_date example is not very robust, since the pub_date will tend to be reset every time the site is rebuilt.  Can you suggest other use cases?

(I hate having to manually enter publication and last-updated dates, too.  My solution was the [lektor-git-timestamp][] plugin.)

[lektor-git-timestamp]: <https://pypi.org/project/lektor-git-timestamp/>

**Comment by dwt** [comment]
Hm, I haven't tested enough probably. But I was Hoping that the default value gets written into the model and thus doesn't change after page creation.

**Comment by dairiki** [comment]
> Hm, I haven't tested enough probably. But I was Hoping that the default value gets written into the model and thus doesn't change after page creation.

Nope.  Whenever the page gets rebuilt, the default will be recomputed.  (Due to Lektor's use of the build cache, the page may not get rebuilt every time the site is rebuilt — but the computed default won't get stored anywhere other than the build cache.)

**Comment by dwt** [comment]
> Nope. Whenever the page gets rebuilt, the default will be recomputed. (Due to Lektor's use of the build cache, the page may not get rebuilt every time the site is rebuilt — but the computed default won't get stored anywhere other than the build cache.)
> 
:-( That sucks.

May I then promote that it would be a really useful to have a feature where one can provide a default value for a model value, that is able to accommodate a) static values so that I can pre-fill my Writer name into the Author field, without precluding my inviting someone else and properly attributing him. Also b) this should be powerful enough that a template expression like `foo {{ date.today() }} bar` can be used to mix in dynamic content and stuff looked up from a database (for example).

That seems really useful - and something I would actually have expected the `default` field in the field description to stand for.

Can you give a hint why it is the way it is currently? It kind of seems counterintuitive that you can set a `default` - but then that default is not actually … used?

**Comment by dairiki** [comment]
Field default values do, of course, get used.  Currently, they are not "dynamic" — that is they are not jinja-evaluated — though, as noted above, a custom type could always change that.  (Changing that for all/existing types would be, I think, confusing, and backward-incompatible, in any case.)

A default author will work fine.  E.g., if your `models/page.ini` includes something like
```
[fields.author]
label = Author
type = string
default = Joe Bloe
```
then any page which does not define a value for the `author` field in its `contents.lr` will get `this.author = "Joe Bloe"`.  (However, note that if you, at a later time, change the default in `models/page.ini`, that will change the author for all pages which do not explicitly specify their own author in their `contents.lr` file.)

**Comment by dwt** [comment]
@dairiki Thank you for the explanation. Do you happen to know why the default value doesn't actually get written to the Model?

An maybe a slightly different angle: What about a `dynamic_default` value which is dynamic (i.e. allows jinja expressions) and which does get saved to the model, so it allows to set a default value for fields like `publish_date`? I could imagine that over time we could deprecate `default` and after some time where it emits warnings have `default`  behave the same as `dynamic_default`?

**Comment by dwt** [comment]
Because if I understand you right, even overriding the `Type.value_from_raw_with_default()` would not save the value in the model file?

**Comment by yagebu** [comment]
I think the current behaviour of `default` makes sense as it allows you to not having to write that field to every page using that model and I don't think we should change it. Allowing this to be dynamic would maybe work for some cases but have a hard-to-explain behaviour due to it probably inconsistently updating thanks to the build cache.

I think your usecase should be accomodated by a new option for the field or model to allow you specify some initial values if you create a page with that model in the admin UI (that should cover it, right?)

**Comment by dwt** [comment]
@yagebu yes, I think `initial_value` or `dynamic_initial_value` would be great. I still think that in the long run maybe `default` should be changed to behave this way - but that can wait.

**Comment by dairiki** [comment]
More proposed use cases might help clarify the issue here.

If all you really want is `pub_date` (and maybe `mod_date`), then perhaps the GUI should just explicitly add a `_ctime` and `_mtime` field to the records when they are edited.  Note that `dynamic_initial_value` doesn't help with a `mod_date`.

Also note that, by design, the admin GUI is, essentially, just an add-on tool.  It is entirely possible to build a Lektor website from scratch without using the admin GUI — i.e. the `contents.lr` and other files in the `contents` tree can be generated by hand-editing or some other method altogether.  (So features like this which only affect the admin GUI will only work when the site is edited using the admin GUI.  This is admittedly the common use case, but it's not the only one.)
